bfvmm/src/hve/arch/intel_x64/ept/memory_map.cpp

```diff
@@ -0,0 +1,372 @@
+            return this->map_pte_to_page(gpa, hpa);
+
+        default:
+            throw std::logic_error("map: invalid ept page size specified");
+    }
+}
+
+void
+memory_map::unmap(gpa_t gpa)
+{
+    auto leaf = gpa_to_epte(gpa);
+    epte::clear(leaf);
+}
+
+epte_t &
+memory_map::gpa_to_epte(gpa_t gpa)
+{
+    auto pml4e = this->gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) {
+        throw std::runtime_error("gpa_to_epte: failed to resolve gpa->epte, "
+            "gpa is not mapped at the 512GB level");
+    }
+
+    auto &pdpte = this->gpa_to_pdpte(gpa, pml4e);
+    if (!epte::is_present(pdpte)) {
+        throw std::runtime_error("gpa_to_epte: failed to resolve gpa->epte, "
+            "gpa is not mapped at the 1GB level");
+    }
+    if (epte::is_leaf_entry(pdpte)) {
+        return pdpte;
+    }
+
+    auto &pde = this->gpa_to_pde(gpa, pdpte);
+    if (!epte::is_present(pde)) {
+        throw std::runtime_error("gpa_to_epte: failed to resolve gpa->epte, "
+            "gpa is not mapped at the 2MB level");
+    }
+    if (epte::is_leaf_entry(pde)) {
+        return pde;
+    }
+
+    auto &pte = this->gpa_to_pte(gpa, pde);
+    if (!epte::is_present(pte)) {
+        throw std::runtime_error("gpa_to_epte: failed to resolve gpa->epte, "
+            "gpa is not mapped at the 4KB level");
+    }
+    if (epte::is_leaf_entry(pte)) {
+        return pte;
+    }
+
+    throw std::runtime_error("gpa_to_epte: extended page tables corrupted");
+}
+
+hpa_t
+memory_map::gpa_to_hpa(gpa_t gpa)
+{
+    auto &pml4e = this->gpa_to_pml4e(gpa);
+    if (!epte::is_present(pml4e)) {
+        throw std::runtime_error("gpa_to_hpa: failed to resolve gpa->epte, gpa "
+            "is not mapped at the 512GB level");
+    }
+
+    auto &pdpte = this->gpa_to_pdpte(gpa, pml4e);
+    if (!epte::is_present(pdpte)) {
+        throw std::runtime_error("gpa_to_hpa: failed to resolve gpa->epte, gpa "
+            "is not mapped at the 1GB level");
+    }
+    if (epte::is_leaf_entry(pdpte)) {
+        return pdpte::page_address::get_effective_address(pdpte, gpa);
+    }
+
+    auto &pde = this->gpa_to_pde(gpa, pdpte);
+    if (!epte::is_present(pde)) {
+        throw std::runtime_error("gpa_to_hpa: failed to resolve gpa->epte, gpa "
+            "is not mapped at the 2MB level");
+    }
+    if (epte::is_leaf_entry(pde)) {
+        return pde::page_address::get_effective_address(pde, gpa);
+    }
+
+    auto &pte = this->gpa_to_pte(gpa, pde);
+    if (!epte::is_present(pte)) {
+        throw std::runtime_error("gpa_to_hpa: failed to resolve gpa->epte, gpa "
+            "is not mapped at the 4KB level");
+    }
+    if (epte::is_leaf_entry(pte)) {
+        return pte::page_address::get_effective_address(pte, gpa);
+    }
+
+    throw std::runtime_error("gpa_to_hpa: extended page tables corrupted");
+}
+
+std::vector<memory_descriptor>
+memory_map::to_mdl() const
+{
+    std::vector<memory_descriptor> mdl{0};
```

Comment:
don't initialize the vector. This is a class that has it's own default initializer